File: src/zig/math/include/quat.zig
const vec = @import("vec.zig");
const mat = @import("mat.zig");

// Quaternion type using Vec3 for vector component and scalar for w
pub const Quat = extern struct {
    v: vec.Vec3, // Vector part (x, y, z)
    w: f32, // Scalar part

    pub inline fn asArray(self: *const Quat) *const [4]f32 {
        return @ptrCast(&self.v.x);
    }

    pub inline fn fromArray(arr: *const [4]f32) Quat {
        return Quat{ .v = vec.Vec3{ .x = arr[0], .y = arr[1], .z = arr[2] }, .w = arr[3] };
    }
};

// Constructor functions
pub extern fn q(x: f32, y: f32, z: f32, w: f32) Quat;
pub extern fn qFromVec(v: vec.Vec3, w: f32) Quat;
pub extern fn qId() Quat;
pub extern fn qZero() Quat;

// Create quaternion from axis-angle representation
pub extern fn qAxis(axis: vec.Vec3, angle: f32) Quat;

// Create quaternion from Euler angles (in radians)
pub extern fn qEuler(x: f32, y: f32, z: f32) Quat;

// Create quaternion from rotation matrix
pub extern fn qFromM3(m: mat.Mat3) Quat;
pub extern fn qFromM4(m: mat.Mat4) Quat;

// Convert quaternion to matrices
pub extern fn qToM3(q: Quat) mat.Mat3;
pub extern fn qToM4(q: Quat) mat.Mat4;

// Basic quaternion operations
pub extern fn qAdd(a: Quat, b: Quat) Quat;
pub extern fn qSub(a: Quat, b: Quat) Quat;
pub extern fn qMul(a: Quat, b: Quat) Quat; // Quaternion multiplication
pub extern fn qScale(q: Quat, s: f32) Quat;
pub extern fn qNeg(q: Quat) Quat;

// Quaternion properties
pub extern fn qLen(q: Quat) f32;
pub extern fn qLen2(q: Quat) f32;
pub extern fn qNorm(q: Quat) Quat; // Normalize quaternion
pub extern fn qConj(q: Quat) Quat; // Conjugate
pub extern fn qInv(q: Quat) Quat; // Inverse quaternion

// Dot product and equality check
pub extern fn qDot(a: Quat, b: Quat) f32;
pub extern fn qEq(a: Quat, b: Quat, eps: f32) bool;

// Rotation operations
pub extern fn qRotV3(q: Quat, v: vec.Vec3) vec.Vec3; // Rotate a vector by quaternion

// Interpolation
pub extern fn qLerp(a: Quat, b: Quat, t: f32) Quat; // Linear interpolation
pub extern fn qSlerp(a: Quat, b: Quat, t: f32) Quat; // Spherical linear interpolation
pub extern fn qNlerp(a: Quat, b: Quat, t: f32) Quat; // Normalized linear interpolation

// Extract information
pub extern fn qGetAxis(q: Quat, axis: *vec.Vec3, angle: *f32) void;
pub extern fn qToEuler(q: Quat) vec.Vec3; // Returns Euler angles as Vec3

// Basic rotation quaternions
pub extern fn qRotX(angle: f32) Quat;
pub extern fn qRotY(angle: f32) Quat;
pub extern fn qRotZ(angle: f32) Quat;

pub extern fn qRoll(quat: Quat) f32;
pub extern fn qPitch(quat: Quat) f32;
pub extern fn qYaw(quat: Quat) f32;

// Additional utility functions
pub extern fn qLookAt(dir: vec.Vec3, up: vec.Vec3) Quat; // Quaternion that rotates towards a direction
pub extern fn qFromTo(from: vec.Vec3, to: vec.Vec3) Quat; // Quaternion to rotate from one vector to another

// Vector and scalar extraction
pub extern fn qGetVec(q: Quat) vec.Vec3; // Get vector part
pub extern fn qGetW(q: Quat) f32; // Get scalar part

File: old/2025/math/bindings/macos/Quat.swift
// /Users/solmidnight/work/angelite/src/swift/gfx/Quat.swift
// Generated binding for quat.zig

import Foundation

@frozen public struct Quat {
  public var v: Vec3
  public var w: Float

  public init(v: Vec3, w: Float) {
    self.v = v
    self.w = w
  }
}

@_silgen_name("q")
public func q(x: Float, y: Float, z: Float, w: Float) -> Quat

@_silgen_name("qFromVec")
public func qFromVec(v: Vec3, w: Float) -> Quat

@_silgen_name("qId")
public func qId() -> Quat

@_silgen_name("qZero")
public func qZero() -> Quat

@_silgen_name("qAxis")
public func qAxis(axis: Vec3, angle: Float) -> Quat

@_silgen_name("qEuler")
public func qEuler(x: Float, y: Float, z: Float) -> Quat

@_silgen_name("qFromM3")
public func qFromM3(m: Mat3) -> Quat

@_silgen_name("qFromM4")
public func qFromM4(m: Mat4) -> Quat

@_silgen_name("qToM3")
public func qToM3(q: Quat) -> Mat3

@_silgen_name("qToM4")
public func qToM4(q: Quat) -> Mat4

@_silgen_name("qAdd")
public func qAdd(a: Quat, b: Quat) -> Quat

@_silgen_name("qSub")
public func qSub(a: Quat, b: Quat) -> Quat

@_silgen_name("qMul")
public func qMul(a: Quat, b: Quat) -> Quat

@_silgen_name("qScale")
public func qScale(q: Quat, s: Float) -> Quat

@_silgen_name("qNeg")
public func qNeg(q: Quat) -> Quat

@_silgen_name("qLen")
public func qLen(q: Quat) -> Float

@_silgen_name("qLen2")
public func qLen2(q: Quat) -> Float

@_silgen_name("qNorm")
public func qNorm(q: Quat) -> Quat

@_silgen_name("qConj")
public func qConj(q: Quat) -> Quat

@_silgen_name("qInv")
public func qInv(q: Quat) -> Quat

@_silgen_name("qDot")
public func qDot(a: Quat, b: Quat) -> Float

@_silgen_name("qEq")
public func qEq(a: Quat, b: Quat, eps: Float) -> Bool

@_silgen_name("qRotV3")
public func qRotV3(q: Quat, v: Vec3) -> Vec3

@_silgen_name("qLerp")
public func qLerp(a: Quat, b: Quat, t: Float) -> Quat

@_silgen_name("qSlerp")
public func qSlerp(a: Quat, b: Quat, t: Float) -> Quat

@_silgen_name("qNlerp")
public func qNlerp(a: Quat, b: Quat, t: Float) -> Quat

@_silgen_name("qGetAxis")
public func qGetAxis(
  q: Quat, axis: UnsafeMutablePointer<Vec3>?, angle: UnsafeMutablePointer<Float>?)

@_silgen_name("qToEuler")
public func qToEuler(q: Quat) -> Vec3

@_silgen_name("qRotX")
public func qRotX(angle: Float) -> Quat

@_silgen_name("qRotY")
public func qRotY(angle: Float) -> Quat

@_silgen_name("qRotZ")
public func qRotZ(angle: Float) -> Quat

@_silgen_name("qRoll")
public func qRoll(quat: Quat) -> Float

@_silgen_name("qPitch")
public func qPitch(quat: Quat) -> Float

@_silgen_name("qYaw")
public func qYaw(quat: Quat) -> Float

@_silgen_name("qLookAt")
public func qLookAt(dir: Vec3, up: Vec3) -> Quat

@_silgen_name("qFromTo")
public func qFromTo(from: Vec3, to: Vec3) -> Quat

@_silgen_name("qGetVec")
public func qGetVec(q: Quat) -> Vec3

@_silgen_name("qGetW")
public func qGetW(q: Quat) -> Float
